### PR TITLE
Integrate WASI test suite

### PR DESCRIPTION
**Pull request by eloparco**
Add [wasi-testsuite](https://github.com/WebAssembly/wasi-testsuite) to set of tests executed from `test_wamr.sh`. 
Additional description here: #1761

### CODE REVIEW

**Comment by eloparco** [comment]
What's missing in the PR (I'll resume the work after I return from vacation, but I pushed the change to get some feedback):
- is the `return 0` [here](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/product-mini/platforms/posix/main.c#L701) actually needed? I commented it out and ran the spec test in GitHub actions and they succeed without that part; I asked already in #1377
- [one test](https://github.com/WebAssembly/wasi-testsuite/blob/main/tests/c/testsuite/stat-dev-ino.c) fails right now, I need to look into it

File: product-mini/platforms/posix/main.c [hunk 1]
```diff
@@ -696,10 +696,10 @@ main(int argc, char *argv[])
...
-    (void)ret;
-    return 0;
-#else
+    // #if WASM_ENABLE_SPEC_TEST != 0
+    //     (void)ret;
+    //     return 0;
+    // #else
     return ret;
-#endif
+    // #endif
```
Comment (yamt) [inline product-mini/platforms/posix/main.c:704]: how is this related to the rest of the PR?

File: tests/wamr-test-suites/test_wamr.sh [hunk 2]
```diff
@@ -453,7 +453,7 @@ function spec_test()
     echo -e "\nFinish spec tests" | tee -a ${REPORT_DIR}/spec_test_report.txt
 }
 
-function wasi_test()
+function wasi_test_OUTDATED()
```
Comment (yamt) [inline tests/wamr-test-suites/test_wamr.sh:456]: at this point it has very different coverage to simply replace.

File: product-mini/platforms/posix/main.c (hunk 1 above)
Comment (eloparco) [reply product-mini/platforms/posix/main.c:704]: In [wasi-testsuite](https://github.com/WebAssembly/wasi-testsuite) there are some tests that check the return code and, because of the hardcoded `return 0`, would fail.
It has been fixed in the `main` branch now #1377 #1839.

To merge this PR, we need https://github.com/WebAssembly/wasi-testsuite/pull/40 to be merged and `main` to be merged into [dev/wasi_threads](https://github.com/bytecodealliance/wasm-micro-runtime/tree/dev/wasi_threads).

Comment (loganek) [reply product-mini/platforms/posix/main.c:704]: @wenyongh would you be able to help with the second point and merge `main` to `dev/wasi_threads` and unblock this PR?

File: tests/wamr-test-suites/test_wamr.sh (hunk 2 above)
Comment (loganek) [reply tests/wamr-test-suites/test_wamr.sh:456]: Agree, I'd rather keep this name unchanged, and maybe find another name for the new test suite (`wasi_certification_tests()`?). Once we have the same coverage, we can delete this test suite completely.

File: tests/wamr-test-suites/test_wamr.sh [hunk 3]
```diff
@@ -469,6 +469,37 @@ function wasi_test()
...
+    else
+        echo "wasi exists, pulling updates from github"
+        pushd wasi-testsuite
+        git fetch
+        git reset --hard origin/prod/testsuite-base
+        popd
+    fi
+
+    cd wasi-testsuite
+    python3 -m pip install -r test-runner/requirements.txt
```
Comment (loganek) [inline tests/wamr-test-suites/test_wamr.sh:490]: I wonder if we can do it in virtual environment [1] so we don't pollute host's environment with packages that they might not want to have? 

[1] https://docs.python.org/3/tutorial/venv.html

File: tests/wamr-test-suites/test_wamr.sh [hunk 4]
```diff
@@ -469,6 +469,37 @@ function wasi_test()
...
+    IWASM_PATH=$(dirname ${IWASM_CMD})
+    PATH=${PATH}:${IWASM_PATH} python3 test-runner/wasi_test_runner.py \
+                -r adapters/wasm-micro-runtime.sh \
+                -t \
+                    tests/c/testsuite/ \
+                    tests/assemblyscript/testsuite/ \
+                | tee -a ${REPORT_DIR}/wasi_test_report.txt
+    [[ ${PIPESTATUS[0]} -ne 0 ]] && exit 1
```
Comment (loganek) [inline tests/wamr-test-suites/test_wamr.sh:498]: does it mean that in case of failure the L500 won't be executed? Is that desired? If we don't have the last line, we probably wouldn't know from this log file whether the test execution completed or just broke half way through.

Comment (eloparco) [reply tests/wamr-test-suites/test_wamr.sh:498]: I was replicating what's done already for spec tests https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/tests/wamr-test-suites/test_wamr.sh#L450. I'll update it in both cases to add something to the log file before exiting.

File: product-mini/platforms/posix/main.c (hunk 1 above)
Comment (wenyongh) [reply product-mini/platforms/posix/main.c:704]: @eloparco @loganek I have merged `main` into `dev/wasi_threads`, since there are code conflicts, I applied the merging manually, please check whether it works for you:
 https://github.com/bytecodealliance/wasm-micro-runtime/commit/3403f0a7f4d838f260b55b327da8be00a8644229

Comment (loganek) [reply product-mini/platforms/posix/main.c:704]: @wenyongh thanks a lot, do you remember which files required manual conflict resolution?

File: tests/wamr-test-suites/test_wamr.sh (hunk 2 above)
Comment (eloparco) [reply tests/wamr-test-suites/test_wamr.sh:456]: and for the new `wasi_certification_tests`, should I just skip the coverage generation? I don't think we can use the same (`lcov`) as in the other tests?

File: product-mini/platforms/posix/main.c (hunk 1 above)
Comment (wenyongh) [reply product-mini/platforms/posix/main.c:704]: @loganek Not the files related to wasi-threads, but some files modified in main branch. I used mode `rebase and merge` instead of mode `merge` to merge `main` into `dev/wasi_threads` last time, it didn't produce "a new generated commit" and seems that `dev/wasi_threads` didn't remember the sync-up status from `main`, so when the same line of the same file was modified in main branch by new commit, conflict occurs. Similar conflicts occurred when I tried merging main into `dev/fast_jit_simd` and `gitbook`, I think I should use `merge` mode to merge main into other branches next time.

Comment (loganek) [reply product-mini/platforms/posix/main.c:704]: ok, thanks. Looks good to me. I think once we're ready with the dev branch, I'll rebase it completely on top of main and open a final PR to rebase-merge it, to keep the git tree clean.

File: tests/wamr-test-suites/test_wamr.sh (hunk 2 above)
Comment (loganek) [reply tests/wamr-test-suites/test_wamr.sh:456]: I'd keep the integration simple for now, and we can add more features in the future if needed

File: tests/wamr-test-suites/test_wamr.sh [hunk 5]
```diff
@@ -469,6 +472,43 @@ function wasi_test()
...
+    python3 -m pip install -r test-runner/requirements.txt
+    IWASM_PATH=$(dirname ${IWASM_CMD})
+    PATH=${PATH}:${IWASM_PATH} python3 test-runner/wasi_test_runner.py \
+                -r adapters/wasm-micro-runtime.sh \
+                -t \
+                    tests/c/testsuite/ \
+                    tests/assemblyscript/testsuite/ \
+                | tee -a ${REPORT_DIR}/wasi_test_report.txt
+    exit_code=${PIPESTATUS[0]}
+    deactivate
```
Comment (loganek) [inline tests/wamr-test-suites/test_wamr.sh:503]: what's that for?

Comment (eloparco) [reply tests/wamr-test-suites/test_wamr.sh:503]: to stop using the `wasi-env` virtual environment created before; if another test suite has dependencies it will create another virtual environment with its own dependencies

Comment (loganek) [reply tests/wamr-test-suites/test_wamr.sh:503]: Ah right, thanks.

File: .github/workflows/compilation_on_android_ubuntu.yml [hunk 6]
```diff
@@ -475,7 +477,7 @@ jobs:
           sudo apt-get update &&
           sudo apt install -y g++-multilib lib32gcc-9-dev
 
-      - name: run spec tests x86_32
-        if: env.TEST_ON_X86_32 == 'true'
+      - name: run tests x86_32
+        if: env.TEST_ON_X86_32 == 'true' && matrix.test_option != '$WASI_TEST_OPTIONS'
```
Comment (loganek) [inline .github/workflows/compilation_on_android_ubuntu.yml:481]: What's the reason for not running wasi tests on x86-32?

**Comment by loganek** [comment]
I don't think the PR has anything to do with wasi threads, could you change the destination branch to `main`?

File: tests/wamr-test-suites/test_wamr.sh [hunk 7]
```diff
@@ -469,6 +472,43 @@ function wasi_test()
...
+    cd ${WORK_DIR}
+    if [ ! -d "wasi-testsuite" ]; then
+        echo "wasi not exist, clone it from github"
+        git clone -b prod/testsuite-base \
+            --single-branch https://github.com/WebAssembly/wasi-testsuite.git
+    else
+        echo "wasi exists, pulling updates from github"
+        pushd wasi-testsuite
+        git fetch
+        git reset --hard origin/prod/testsuite-base
```
Comment (loganek) [inline tests/wamr-test-suites/test_wamr.sh:488]: I'd suggest at least for now we point to a specific commit. Using always the latest testsuite is a bit risky and might block the pipeline, especially that there's a lot going on in the wasi-testsuite repo now. We might want to choose to use always the latest version in the future, but I'd point to specific commit for now.

File: .github/workflows/compilation_on_android_ubuntu.yml (hunk 6 above)
Comment (loganek) [reply .github/workflows/compilation_on_android_ubuntu.yml:481]: I think it'd be worth running those tests on x86-32 as well (to make sure WASI implementation works on those platforms as well).

Comment (wenyongh) [reply .github/workflows/compilation_on_android_ubuntu.yml]: Do we want to run wasi tests on x86-32? For the output of the CI, I found that x86-32 wasi tests are ignored:
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/3834321537/jobs/6526922968
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/3834321537/jobs/6526923240

Seems that we need to change the below lines in the CI file
```yml
- name: set env variable(if x86_32 test needed)
        if: >
          (matrix.test_option == '$DEFAULT_TEST_OPTIONS' || matrix.test_option == '$THREADS_TEST_OPTIONS')
          && matrix.running_mode != 'fast-jit' && matrix.running_mode != 'jit'
        run: echo "TEST_ON_X86_32=true" >> $GITHUB_ENV
```
into:
```yml
- name: set env variable(if x86_32 test needed)
        if: >
          (matrix.test_option == '$DEFAULT_TEST_OPTIONS' || matrix.test_option == '$THREADS_TEST_OPTIONS'
           || matrix.test_option == '$WASI_TEST_OPTIONS')
          && matrix.running_mode != 'fast-jit' && matrix.running_mode != 'jit'
        run: echo "TEST_ON_X86_32=true" >> $GITHUB_ENV
```

**Comment by lum1n0us** [comment]
LGTM

File: tests/wamr-test-suites/test_wamr.sh (hunk 2 above)
Comment (loganek) [reply tests/wamr-test-suites/test_wamr.sh]: It's not an official concept (at least not yet), but we strive to provide an extensive coverage so the testsuite can act as certification tests.

**Review by loganek — APPROVED** [review]
LGTM
